superbench/tools/system_info.py

```diff
@@ -0,0 +1,442 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+"""Generate system config."""
+
+import json
+import subprocess
+import xmltodict
+from pathlib import Path
+
+
+class SystemInfo():
+    """Systsem info class."""
+    def __run_cmd(self, command, non_root=False):
+        """Run the command in root right or non-root right and return the stdout string.
+
+        Args:
+            command (string): the command to run in terminal.
+            non_root (bool): the mode to run the command, root or non-root.
+
+        Returns:
+            string: the stdout string of the command.
+        """
+        prefix = ''
+        if not non_root:
+            prefix = 'sudo '
+        output = subprocess.run(
+            prefix + command,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            shell=True,
+            check=False,
+            universal_newlines=True
+        )
+        return output.stdout
+
+    def __cal_indent(self, line, symbol='\t'):
+        r"""Calcudate the indent count of the symbol in the beginning of the line.
+
+        Args:
+            line (string): the line to calculate the indent count.
+            symbol (str, optional): the symbol of the indent. Defaults to '\t'.
+
+        Returns:
+            int: the indent count of the symbol in the beginning of the line.
+        """
+        count = 0
+        for char in line:
+            if char == symbol:
+                count += 1
+            else:
+                break
+        return count
+
+    def __parse_key_value_lines(self, lines, selected=None, omitted_value=None, symbol=':'):    # noqa: C901
+        """Parse the lines like "key:value" and convert them to dict.
+
+        If selected is None, parse all lines, otherwise only parse the lines which contain selected.
+        If omitted_value is None, accept any value in dict, otherwise drop the item whose value in omitted_value.
+
+        Args:
+            lines (list): the lines to parse.
+            selected (list, optional): list of select keys. Defaults to None.
+            omitted_value (list, optional): list of omitted values. Defaults to None.
+
+        Returns:
+            dict: the result in dict.
+        """
+        dict = {}
+        key = ''
+        value = ''
+        i = 0
+        length = len(lines)
+        while i < length:
+            line = lines[i]
+            is_selected = True
+            if selected is not None:
+                is_selected = False
+                for key in selected:
+                    if key in line:
+                        is_selected = True
+            if not is_selected:
+                i += 1
+                continue
+            # process with indent recursively
```

Comment:
Cannot understand the logic clearly. Would you please give an example for this recursive parsing?

Reply:
I will do the recursive parsing to deal with the different indent in content into different level dict, for example, the content is as the following, if next line has more indent compared to current line, I will keep this line as a key, and then recursively parse the lines which have the same indent with next line as the value of the key.
```
CA 'mlx5_0'
        CA type: MT4123
        Number of ports: 1
        Firmware version: 20.30.1004
        Hardware version: 0
        Node GUID: 0x88e9a4ffff0a28f4
        System image GUID: 0x88e9a4ffff0a28f4
        Port 1:
                State: Down
                Physical state: Disabled
                Rate: 40
                Base lid: 0
                LMC: 0
                SM lid: 0
                Capability mask: 0x00010000
                Port GUID: 0x8ae9a4fffe0a28f4
                Link layer: Ethernet
CA 'mlx5_1'
        CA type: MT4123
        Number of ports: 1
        Firmware version: 20.30.1004
        Hardware version: 0
        Node GUID: 0x88e9a4ffff0a28f5
        System image GUID: 0x88e9a4ffff0a28f4
        Port 1:
                State: Active
                Physical state: LinkUp
                Rate: 200
                Base lid: 38
                LMC: 0
                SM lid: 1
                Capability mask: 0x2651e848
                Port GUID: 0x88e9a4ffff0a28f5
                Link layer: InfiniBand
```
so the result of the parsing will be like as the following:
```
                "CA 'mlx5_0'": {
                    "CA type": "MT4123",
                    "Number of ports": "1",
                    "Firmware version": "20.30.1004",
                    "Hardware version": "0",
                    "Node GUID": "0x88e9a4ffff0a28f4",
                    "System image GUID": "0x88e9a4ffff0a28f4",
                    "Port 1:": {
                        "State": "Down",
                        "Physical state": "Disabled",
                        "Rate": "40",
                        "Base lid": "0",
                        "LMC": "0",
                        "SM lid": "0",
                        "Capability mask": "0x00010000",
                        "Port GUID": "0x8ae9a4fffe0a28f4",
                        "Link layer": "Ethernet"
                    }
                },
                "CA 'mlx5_1'": {
                    "CA type": "MT4123",
                    "Number of ports": "1",
                    "Firmware version": "20.30.1004",
                    "Hardware version": "0",
                    "Node GUID": "0x88e9a4ffff0a28f5",
                    "System image GUID": "0x88e9a4ffff0a28f4",
                    "Port 1:": {
                        "State": "Active",
                        "Physical state": "LinkUp",
                        "Rate": "200",
                        "Base lid": "38",
                        "LMC": "0",
                        "SM lid": "1",
                        "Capability mask": "0x2651e848",
                        "Port GUID": "0x88e9a4ffff0a28f5",
                        "Link layer": "InfiniBand"
                    }
                },
```